File: 0107-binary-tree-level-order-traversal-ii/0107-binary-tree-level-order-traversal-ii.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def levelOrderBottom(self, root: Optional[TreeNode]) -> List[List[int]]:
        def bfs(node):
            if not node:
                return []
            
            queue = deque([node])
            result = []

            while queue:
                temp = []
                size = len(queue)
                for _ in range(size):
                    node = queue.popleft()
                    temp.append(node.val)

                    if node.left:
                        queue.append(node.left)
                    if node.right:
                        queue.append(node.right)
                result.append(temp)
            return result[::-1]
    
        return bfs(root)

File: 0107-binary-tree-level-order-traversal-ii/0107-binary-tree-level-order-traversal-ii.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number[][]}
 */
var levelOrderBottom = function(root) {
    const bfs = (node) => {
        if(!node){
            return []
        }

        const result = []
        const queue = [node]

        while(queue.length > 0){
            const temp = []
            const size = queue.length

            for (let i = 0; i< size; i++){
                const node = queue.shift()
                temp.push(node.val)


                if(node.left){
                    queue.push(node.left)
                }
                if(node.right){
                    queue.push(node.right)
                }

            }
            result.push(temp)
        }
        return result.reverse()
    }

    return bfs(root)
};